File: src/analysis/plot.py
"""Helper functions for transforming and visualizing atlas data for analysis."""

import hydra
import os

import numpy as np
import pandas as pd
import plotnine as pn

from collections import Counter


from sciterra import Atlas, Cartographer
from sciterra.vectorization.vectorizer import Vectorizer
from sciterra.mapping.tracing import search_converged_ids


##############################################################################
# Plotting
##############################################################################


def fields_histogram(atl: Atlas) -> pn.ggplot:
    """Get a (horizontal) historgram with fields on the y-axis, and publication counts on the x axis, corresponding to the distribution of fields of study for publications in the atlas."""
    fields = Counter(
        [field for pub in atl.publications.values() for field in pub.fields_of_study]
    )

    return (
        pn.ggplot(
            data=pd.DataFrame(
                data=[(field, count) for field, count in fields.items()],
                columns=["field", "count"],
            ),
            mapping=pn.aes(x="field", y="count"),
        )
        # + pn.scale_y_continuous(trans = "log10")
        + pn.geom_col()
        + pn.coord_flip()
    )


def cpy_histogram(
    data: pd.DataFrame,
    min_cpy=1,
    max_cpy=100,
) -> pn.ggplot:
    """Get a histogram of citations per year, given a dataframe of measurements."""

    df_f = data[
        (data["citations_per_year"] >= min_cpy)
        & (data["citations_per_year"] <= max_cpy)
    ]

    # Histogram of cpy
    cpy_hist = (
        pn.ggplot(
            # df,
            df_f,
            mapping=pn.aes(x="citations_per_year"),
        )
        + pn.geom_histogram(bins=50)
        + pn.scale_x_log10()
        + pn.xlab("Citations per year")
        + pn.ylab(f"Count, total={len(df_f)}")
        + pn.theme(axis_title=pn.element_text(size=18))
    )

    return cpy_hist


# TODO: refactor w above
def density_histogram(
    data: pd.DataFrame,
    min_d=-np.inf,
    max_d=np.inf,
) -> pn.ggplot:
    """Get a histogram of density values, given a dataframe of measurements."""

    df_f = data[(data["density"] >= min_d) & (data["density"] <= max_d)]

    # Histogram of cpy
    cpy_hist = (
        pn.ggplot(
            # df,
            df_f,
            mapping=pn.aes(x="density"),
        )
        + pn.geom_histogram(bins=50)
        # + pn.scale_x_log10() # probably don't do this!
        + pn.xlab("Density")
        + pn.ylab(f"Count, total={len(df_f)}")
        + pn.theme(axis_title=pn.element_text(size=18))
    )

    return cpy_hist


# TODO: refactor w above
def years_histogram(
    atl: Atlas,
) -> pn.ggplot:
    """Get a histogram of the years of publication."""

    years = Counter([pub.publication_date.year for pub in atl.publications.values()])

    return (
        pn.ggplot(
            data=pd.DataFrame(
                data=[(year, count) for year, count in years.items()],
                columns=["year", "count"],
            ),
            mapping=pn.aes(x="year", y="count"),
        )
        + pn.geom_col()
    )


# R backend
# TODO: consider adding the identity of pubs in the data, so we can visualize where the center is in the distribution.
def call_r_2d_histograms(
    df_fn: str,
    save_dir: str = None,
    **kwargs,
) -> None:
    """Run the backend R script `src/analysis/plot.R`, which created histograms of metrics vs. cpy, filled by density of data points. This plot is not available in plotnine.

    Args:
        df_fn: the absolute path to the CSV containing the measurements resulting from `atlas_to_measurements`.

        save_dir: the absolute path to the directory to save the plots. Default is the parent directory of `df_fn`.
    """
    if save_dir is None:
        save_dir = os.path.dirname(df_fn)

    # Process args
    cml_args = [df_fn, save_dir]
    cml_kwargs = [
        f"--{key} {value}" for key, value in kwargs.items() if value is not None
    ]
    all_cml_args = " ".join(cml_args + cml_kwargs)

    # Run the R script
    chdir = f"cd {hydra.utils.get_original_cwd()}"
    run_rscript = f"Rscript src/analysis/plot.R {all_cml_args}"
    command = f"{chdir}; {run_rscript}"
    os.system(f"echo '{command}'")
    os.system(command)


##############################################################################
# Data transforms
##############################################################################


# TODO: should we move this into a separate file?
def atlas_to_measurements(
    atl: Atlas,
    vectorizer: Vectorizer,
    converged_pub_ids: list[str] = None,
    num_pubs_added: int = 1000,
    kernel_size=16,  # TODO: find a principled way of selecting this value.
    fields_of_study=None,
    max_year: int = 2020,  # consider 2020, since that is when exp growth falls
    min_year: int = 2000,
) -> pd.DataFrame:
    """Compute the density, edginess, and citations per year metrics for each publication in an atlas w.r.t. a vectorizer and convergence configurations, and return the results in a dataframe.

    Args:
        atl: the Atlas to measure

        vectorizer: the Vectorizer to use to compute density and edginess

        converged_pub_ids: the identifiers of publications to consider for analysis. By default is `None`, and will be inferred on the bases of the next two parameters.

        num_pubs_added: propagated to `search_converged_ids`

        kernel_size: propagated to `search_converged_ids`
    """

    if converged_pub_ids is None:
        converged_pub_ids = search_converged_ids(
            atl,
            num_pubs_added=num_pubs_added,
            kernel_size=kernel_size,
        )

    # Optionally filter only to `field_of_study` publications
    if fields_of_study is not None:
        converged_pub_ids = [
            id
            for id in converged_pub_ids
            if any(fld in atl[id].fields_of_study for fld in atl[id].fields_of_study)
        ]

    # Compute density, edginess
    crt = Cartographer(
        vectorizer=vectorizer,
    )
    topo_metrics = ["density", "edginess"]
    measurements = crt.measure_topography(
        atl,
        ids=converged_pub_ids,
        metrics=topo_metrics,
        kernel_size=kernel_size,
    )

    # Count references.
    references = [len(atl[id].references) for id in converged_pub_ids]

    valid_year = lambda year: year < max_year and year > min_year
    # valid_year = lambda year: year == 2010 # robustness check that trends hold for specific years

    # Get citations
    citations_per_year = [
        atl[id].citation_count / (max_year - atl[id].publication_date.year)
        if (
            valid_year(atl[id].publication_date.year)
            and atl[id].citation_count is not None
        )
        else np.nan  # prev set else to 0, but this is less ambiguous
        for id in converged_pub_ids
    ]

    # Get year
    years = [atl[id].publication_date.year for id in converged_pub_ids]

    # Annotate the center (this feels inefficient, but oh well)
    is_center = [identifier == atl.center for identifier in converged_pub_ids]

    if not any(is_center):
        import warnings

        warnings.warn(
            f"The center publication is not within the set of converged publications."
        )

    df = pd.DataFrame(
        measurements,
        columns=topo_metrics,
    )
    df["references"] = references
    df["citations_per_year"] = citations_per_year
    df["is_center"] = is_center
    # Annotate with ids, which can be helpful for copying atlases
    df["identifier"] = converged_pub_ids
    # Annotate with year of publication
    df["year"] = years

    df = df[~np.isinf(df["density"])]  # drop infs which occur for BOW vectorizer
    # TODO what about other very high densities that result from close to 0?

    df.dropna(
        inplace=True,
    )

    print(f"There are {len(df)} total observations after filtering.")

    return df


import scipy
import matplotlib.pyplot as plt
import warnings
import matplotlib.gridspec as gridspec
from matplotlib.figure import Figure
from matplotlib.axes import Axes


def main_trends_mpl(df_plot: pd.DataFrame) -> tuple[Figure, Axes]:
    """Main trend doubleplot with binned density (z-scaled) on x axis and median cpy and log cpy variance on y axis.

    Args:
        df_plot: dataframe containing columns ['density_bin_z', 'cpy_med', 'log_cpy_var', 'field']. This should only contain measurements filtered to a specific vectorizer, and already filtered between reasonable z-scores (e.g., -3 and 3).

    Returns:
        a pair (fig, ax) corresponding to matplotlib Figure and Axes for the plot.

    """

    rows = ["median", "variance"]
    n_rows = 2

    facecolor = np.array([235, 235, 235]) / 256.0
    fig = plt.figure(figsize=(12, 5 * n_rows), facecolor=facecolor)

    gs = gridspec.GridSpec(n_rows, 1)
    gs.update(hspace=0.05, wspace=0.001)

    x_variable = "density_bin_z"
    y_variable_maps = {
        "median": "log_cpy_mean_z",
        "variance": "log_cpy_std_z",
    }

    # For each trend
    for row_idx, row in enumerate(rows):
        with warnings.catch_warnings():
            warnings.simplefilter("ignore")
            ax = plt.subplot(
                gs[row_idx, 0],
            )

        y_variable = y_variable_maps[row]
        x_observations = df_plot[x_variable]
        y_observations = df_plot[y_variable]

        # N.B.: global scatter, median, percentiles not really well formed across fields, since citation rates are field specific.

        # GLobal

        bin_edges = np.linspace(
            x_observations.min(),
            x_observations.max(),
            10,
        )

        xs = bin_edges[:-1] + 0.33 * (bin_edges[1] - bin_edges[0])

        #######################################
        # Fields
        #######################################

        fields = df_plot.field.unique()
        cmap = plt.cm.jet  # define the colormap

        for field in fields:
            df_field = df_plot[df_plot["field"] == field]

            x_observations_field = df_field[x_variable]
            y_observations_field = df_field[y_variable]

            bin_edges = np.linspace(
                x_observations_field.min(),
                x_observations_field.max(),
                6,
            )

            xs = bin_edges[1:]

            # Binned median
            median, bin_edges, _ = scipy.stats.binned_statistic(
                x_observations_field, y_observations_field, "median", bins=bin_edges
            )
            upper_fn = lambda y: np.nanpercentile(y, 68)
            upper, _, _ = scipy.stats.binned_statistic(
                x_observations_field, y_observations_field, upper_fn, bins=bin_edges
            )
            lower_fn = lambda y: np.nanpercentile(y, 32)
            lower, _, _ = scipy.stats.binned_statistic(
                x_observations_field, y_observations_field, lower_fn, bins=bin_edges
            )
            ax.fill_between(
                xs,
                lower,
                upper,
                color="gray",
                alpha=0.15,
            )

            ax.scatter(
                x_observations_field,
                y_observations_field,
                alpha=0.4,
                s=6,
            )

            ax.plot(
                xs,
                median,
                linewidth=4,
                label=field,
            )

        #######################################
        # Label rows
        #######################################
        if row == "median":
            # ax.set_ylim(0,11)
            ax.set_xlim(-2.75, 2.75)

            ax.set_ylabel(r"Mean, $\mu_{\log(cpy)}$", fontsize=16)

        if row == "variance":
            # ax.set_ylim(0.1, 0.8,)
            ax.set_xlim(-2.75, 2.75)

            ax.set_ylabel(r"Std, $\sigma_{\log(cpy)}$", fontsize=16)

        # Customize ticks
        ax.tick_params(right=True, labelright=True)
        if not ax.get_subplotspec().is_last_row():
            ax.tick_params(axis="x", labelbottom=False)

    axbox = ax.get_position()
    ax.legend(
        prop={
            "size": 14,
        },
        ncol=1,
        # loc = 'upper right',
        loc=(axbox.x0 + -0.12, axbox.y0 + 1.23),
        # loc = (axbox.x0 + 0.6, axbox.y0 + 1.23),
        framealpha=0.5,
    )

    ax.set_xlabel(
        r"Density (z-score) of previously existing research, $\rho$", fontsize=16
    )

    # plt.tight_layout()

    return (fig, ax)


import seaborn as sns
import matplotlib.transforms
from matplotlib import patheffects


def summary_violins(
    df_all: pd.DataFrame, categorization: str, quantile: float = 0.25
) -> tuple[Figure, Axes]:
    """Get a multi violin plot, by field or vectorizer.

    Args:
        df_all: all observations, before any transforming.

        categorization: one of {"fields_of_study_0", "vectorizer"}

        quantile: the upper and lower quantile to compare. E.g., 0.5 is median, so lower 50% and 50% will be compared

    Returns:
        a pair (fig, ax) corresponding to matplotlib Figure and Axes for the plot.
    """
    percent = int(100 * quantile)

    # Transform data
    category_names = sorted(df_all[categorization].unique())
    dfs_category_quantiles = []
    for category in category_names:
        df_cat = df_all[df_all[categorization] == category]

        # Create density categories
        upper_quantile = df_cat["density"].quantile(1 - quantile)
        lower_quantile = df_cat["density"].quantile(quantile)
        df_cat["density_cat"] = "center"
        df_cat.loc[
            df_cat["density"] < lower_quantile, "density_cat"
        ] = f"lower {percent}%"
        df_cat.loc[
            df_cat["density"] > upper_quantile, "density_cat"
        ] = f"upper {percent}%"
        df_cat = df_cat.loc[df_cat["density_cat"] != "center"]
        df_cat["density_cat"] = df_cat["density_cat"].astype("category")

        # Annotate category
        df_cat[categorization] = category

        dfs_category_quantiles.append(df_cat)

    # Combine
    df_quantiles = pd.concat(dfs_category_quantiles)
    # Correct duplicates
    df_quantiles = df_quantiles.reset_index()

    # Start building figure
    fig = plt.figure(figsize=(len(category_names) * 2.5, 6))
    ax = plt.gca()
    ax.set_title("Density")

    sns.violinplot(
        ax=ax,
        data=df_quantiles,
        x=categorization,
        y="log_cpy",
        hue="density_cat",
        split=True,
        inner="quart",
        dodge=True,
        gap=0,
    )

    df_by_cat = df_quantiles.groupby(categorization)
    fraction_changes = []
    fraction_std_changes = []
    for i, key in enumerate(category_names):
        # Get the group
        try:
            df_category = df_by_cat.get_group(key)
        except:
            breakpoint()
        n = df_category.shape[0]

        # Median change
        df_category_by_density = df_category.groupby("density_cat")
        med_cpy = 10.0 ** df_category_by_density["log_cpy"].median()
        fraction_change = med_cpy[f"upper {percent}%"] / med_cpy[f"lower {percent}%"]
        fraction_changes.append(fraction_change)
        median_change_str = (
            rf"$c_{{>{percent}}} = " f"{fraction_change:.2f}" rf"c_{{<{percent}}}$"
        )

        # Median change in width
        std_cpy = 10.0 ** df_category_by_density["log_cpy"].std()
        fraction_std_change = (
            std_cpy[f"upper {percent}%"] / std_cpy[f"lower {percent}%"]
        )
        fraction_std_changes.append(fraction_std_change)
        std_change_str = (
            rf"$\sigma_{{>{percent}}} = "
            f"{fraction_std_change:.2f}"
            rf"\sigma_{{<{percent}}}$"
        )

        text = ax.annotate(
            text=f"n={n}\n" + median_change_str + "\n" + std_change_str,
            xy=(i, 1),
            xycoords=matplotlib.transforms.blended_transform_factory(
                ax.transData,
                ax.transAxes,
            ),
            xytext=(0, -5),
            textcoords="offset points",
            ha="center",
            va="top",
            fontsize=14,
        )
        text.set_path_effects(
            [patheffects.Stroke(linewidth=3, foreground="w"), patheffects.Normal()]
        )

    ax.legend(prop=dict(size=14))
    legend = ax.get_legend()
    legend.set_title("Density percentile", prop=dict(size=14))
    legend.set_loc("upper center")
    legend.set_bbox_to_anchor((0.56, 0.84))
    legend.set_alignment("left")

    return (fig, ax)


def tradeoffs_faceted(
    df_plot: pd.DataFrame, risk: str, returns: str, color: str
) -> pn.ggplot:
    df_all = df_plot[df_plot["type"] == "observed"]
    df_dominant = df_plot[df_plot["type"] == "dominant"]

    return (
        pn.ggplot(
            df_all,
            pn.aes(
                x=risk,
                y=returns,
            ),
        )
        + pn.geom_point(
            pn.aes(color=color),
            size=4,
            alpha=0.8,
        )
        # + pn.geom_line(
        #     df_dominant,
        #     color = "black",
        #     size = 2,
        #     # alpha = 0.2,
        #     linetype = "dashed",
        # )
        + pn.facet_wrap("field")
        # + pn.theme_classic()
        + pn.scale_color_continuous("cividis")
        + pn.labs(color="Density, $\\rho$")
        + pn.xlab("Risk, $\sigma_{\log cpy}$")
        + pn.ylab("Return, $\mu_{\log cpy}$")
        + pn.theme(
            # Axis font
            axis_title=pn.element_text(size=24),
            axis_text=pn.element_text(size=18),
            strip_text=pn.element_text(size=18),
        )
    )


# Could refactor with above
def tradeoffs_aggregated(
    df_plot: pd.DataFrame, risk: str, returns: str, color: str
) -> pn.ggplot:
    df_all = df_plot[df_plot["type"] == "observed"]
    df_dominant = df_plot[df_plot["type"] == "dominant"]

    return (
        pn.ggplot(
            df_all,
            pn.aes(
                x=risk,
                y=returns,
                # color="density_bin_z", #
            ),
        )
        + pn.geom_line(
            df_dominant,
            color="black",
            size=4,
            # alpha = 0.2,
            linetype="dashed",
        )
        + pn.geom_point(
            pn.aes(color=color),
            size=6,
            alpha=0.8,
        )
        + pn.scale_color_continuous("cividis")
        + pn.labs(color="Density, $\\rho$")
        + pn.xlab("Risk, $\sigma_{\log cpy}$")
        + pn.ylab("Return, $\mu_{\log cpy}$")
        + pn.theme_classic()
        + pn.theme(
            # Axis font
            axis_title=pn.element_text(size=24),
            axis_text=pn.element_text(size=18),
        )
    )

File: src/analysis/plot.R
# Auxilliary R script to plot data, called by analysis.py. If plotnine ever matches ggplot's energy, this file won't need to exist.

library(argparse)
library(dplyr)
library(ggridges)
library(tidyverse)
library(viridis)
library(DescTools)

# create parser object
parser <- ArgumentParser()

parser$add_argument(
  "data_fn",
  nargs=1, 
  help="Path to CSV file containing atlas topography measurements."
)

parser$add_argument(
  "save_dir",
  nargs=1,
  help="Path to directory to save plots."
)

parser$add_argument(
  "--max_density",
  type="double",
  help="Drop all observations from plotting density for values of density greater than this value."
)

parser$add_argument(
  "--log_cpy",
  nargs=1,
  type="logical",
  help="Whether to plot the y-axis, citations per year, on a log scale.",
  default=FALSE
)

# Optional args for the std filtering
parser$add_argument(
  "--max_cpy_stds",
  nargs=1,
  type="double",
  help="How many standard deviations above the mean to set the max cpy after z-scaling.",
  default=1
)
parser$add_argument(
  "--max_metric_stds",
  nargs=1,
  type="double",
  help="How many standard deviations above the mean to set the max metric value after z-scaling.",
  default=2
)
parser$add_argument(
  "--min_metric_stds",
  nargs=1,
  type="double",
  help="How many standard deviations below the mean to set the max metric value after z-scaling.",
  default=-2
)

args <- parser$parse_args()


df_fn <- args$data_fn
save_dir <- args$save_dir
max_density <- args$max_density
max_cpy_stds <- args$max_cpy_stds
min_metric_stds <- args$min_metric_stds
max_metric_stds <- args$max_metric_stds
log_cpy <- args$log_cpy

# Load data
df <- read_csv(df_fn)

# TODO: if max_density specified, drop rows with greater density values.
# we could groupby and only do this for density, but for consistency we may as well just exclude those values from edginess too
if (is.null(max_density)) {
  max_density <- Inf
}


df <- df %>% filter(
  density <= max_density,
) %>% filter(
  citations_per_year > 0, # We're going to compute log cpys, so need to avoid log0,
)


get_zscaled_filtered <- function(metric) {

  # Filter to 2 stds, for both vars
  metric_z <- paste(metric, "_z", sep="")

  df_z <- df %>% mutate(
    df,
    metric_z = scale(.data[[metric]]),
    cpy_z = scale(citations_per_year),
    references_z = scale(references),
  )
  
  # Add log transform
  df_z <- df_z %>% mutate(
    df_z,
    logcpy = log10( citations_per_year ),
    logcpy_z = log10( cpy_z ),
  )

  # Filter to mean cpys
  # N.B.: sometimes the center will have more than mean!
  # N.B.: physics, but no others, need +1 std for helpful visualization. This should be taken care of by args.
  df_z <- df_z %>%
    filter(
      cpy_z <= max_cpy_stds
    )

  df_zscaled_filtered <- df_z %>% filter(
    (
      metric_z >= min_metric_stds
      & 
      metric_z <= max_metric_stds
    ),
  ) %>% filter(
      references_z <= 3, # might need to outsource to args and config
  )

  return(df_zscaled_filtered)
}

# Plot metric vs. cpy
metric_vs_cpy <- function(metric) {

  df_zf <- get_zscaled_filtered(metric)

  # Get the center to label
  # N.B.: if the python script raised a warning, this dataframe will be empty and no point will be plotted.
  df_center <- df_zf %>% filter(
    is_center == TRUE
  )
  print("DF CENTER:")
  print(df_center)    

  y <- "citations_per_year"
  if (log_cpy) {
    y <- "logcpy"
  }
  
  plot = (
    ggplot(
      df_zf,
      mapping=aes(
        # x=density_z, 
        x=.data[[metric]], # NOTE that we filter by z-scale, but can still plot the orig values.
        # y=citations_per_year
        y=.data[[y]],
      )
    )
    + geom_density_2d_filled(
      contour_var = "ndensity",
      # alpha=0.2,
    )
    + scale_fill_viridis(option = "viridis", discrete = TRUE)
    # + xlab("Density z-scaled")
    + xlab(str_to_title(metric))
    # + ylab("Citations per year")
    + ylab("Citations per year")

    # Local linear regression
    + geom_smooth(color="orange", size=2, method="loess", span=.3)

    # Running binned median
    + stat_summary_bin(fun = "median", geom = "smooth", bins = 10)

    + geom_point(
      alpha=0.05,
      color="white",
      size=1,
    )
    
    # Annotate the center
    + geom_label(
      data=df_center,
      mapping=aes(label="center publication"),
      nudge_y = -.2,
    )
    + geom_point(
      data=df_center,
      # shape=4,
      size=6,
      color="red",
    )
    
    + ylim(0, NA)
    + theme(
      # axis_title_y=element_blank(),
      axis.title=element_text(size=18),
    )
  )
  return(plot)
}

metric_cpy_basic_regression <- function(metric) {

  df_zf <- get_zscaled_filtered(metric)
  y <- "citations_per_year"
  if (log_cpy) {
    y <- "logcpy"
  }  

  plot = (
    ggplot(
      df_zf,
      mapping=aes(
        # x=density_z, 
        x=.data[[metric]], # NOTE that we filter by z-scale, but can still plot the orig values.
        # y=citations_per_year
        y=.data[[y]],
      )
    )
    + xlab(str_to_title(metric))
    # + ylab("Citations per year")
    + ylab("Citations per year")

    # Local linear regression
    + geom_smooth(color="orange", size=3)

    + geom_point(
      alpha=0.1,
      # color="white",
      size=1,
    )
    
    + ylim(0, NA)
    + theme(
      # axis_title_y=element_blank(),
      axis.title=element_text(size=18),
    )
  )
  return(plot)
}

# Ridges
df_zf <- get_zscaled_filtered("density")
df_zf$density_bin <- cut(
  df_zf$density, 
  breaks = seq(
    min(df_zf$density), 
    max(df_zf$density), 
    length.out = 21
  )
)

joyplot = (
  ggplot(
    df_zf,
    aes(
      x = citations_per_year,
      y = density_bin,
      color = density_bin,
      fill = density_bin,
      alpha = 0.2,
    )
  )
  + geom_density_ridges()
  # + xlim(8,10)
  + scale_color_viridis(discrete = TRUE)
  + scale_fill_viridis(discrete = TRUE)
)
save_fn = paste(save_dir, "/", "density_joyplot.png", sep="")
ggsave(
    save_fn,
    plot=joyplot,
    width=10,
    height=10,
)

# Compute entropy for each bin
entropy_estimator <- function(observations) {
  if (length(observations) == 0) {
    return(NaN)
  }
  
  # Calculate bin width for each dimension
  bin_width <- 20 / 21

  # Compute histogram
  # Compute histogram
  # we could use seq(min_value, max_value, length.out = num_bins + 1),
  # but since we essentially filter to <=10 cpys anyway, just estimate this range evenly
  hist <- hist(observations, breaks = 0:20, plot = FALSE)

  # Compute probabilities for each bin
  bin_probabilities <- hist$counts / length(observations)

  # Calculate entropy for this dimension
  entropy_estimate <- -sum(bin_probabilities * log(bin_probabilities + 1e-16) / bin_width)  # Add small value to avoid log(0)

  return(entropy_estimate)
}

# Now we'll filter to less than 20 cpys; the vast majority of data has less than 10 cpys
df_zf <- df_zf %>% filter(
  citations_per_year <= 20,
)

value_counts <-  count(df_zf, density_bin)
density_bin_counts <- value_counts$n
density_levels <- value_counts$density_bin

cpy_ents <- c()
bin_starts <- c()
for (val in density_levels) {
  # need to do some regexing on the density bin name string
  text <- val
  pattern <- "\\(([-+]?[0-9]*\\.?[0-9]+),"
  numbers <- str_extract_all(text, pattern)
  start <- as.double(lapply(numbers[[1]], function(x) gsub("[(,]", "", x)))

  # now filter and compute entropy
  df_val <- df_zf %>% filter(density_bin == val)
  cpy_values <- unlist(c(df_val["citations_per_year"]))
  h <- entropy_estimator(cpy_values)
  cpy_ents <- append(cpy_ents, h)
  bin_starts <- append(bin_starts, start)
}

df_ent <- tibble(cpy_ents, bin_starts, density_bin_counts)
cpy_ent <- (
  ggplot(
    df_ent,
    aes(
      x=bin_starts,      
      y=density_bin_counts,
      fill=cpy_ents,
    )
  )
  # + geom_point(size=10)
  + geom_col()
  + scale_fill_viridis()
  + labs(fill="citation entropy,\nH(CPY | density_bin)\n")
)
save_fn = paste(save_dir, "/", "citation_entropy_vs_density_bins.png", sep="")
ggsave(
    save_fn,
    plot=cpy_ent,
    width=10,
    height=10,
)

# try visualizing distribution of cpy per density differently,
# this is an alternative to the joy plot.

joyalt <- (
  ggplot(
    df_zf,
  )
  + geom_density(
    aes(
      x=citations_per_year,
      # color = density_bin,
      fill = density_bin,
    ),
    alpha=0.75,    
  )
  + scale_fill_viridis( discrete = TRUE)
)
save_fn = paste(save_dir, "/", "joyalt.png", sep="")
ggsave(
    save_fn,
    plot=joyalt,
    width=10,
    height=10,
)

h_regression <- (
    ggplot(
        df_ent, 
        aes(
            x=bin_starts,
            y=cpy_ents,
        )        
    )
    + geom_point(
        aes(
            size=density_bin_counts,
            alpha=density_bin_counts,
        )
    )
    + geom_smooth(method="lm", color="orange")
    + xlab("density bin")
    + ylab("citation entropy, H(CPY | density_bin)\n")
    + labs(size="count")
    + labs(alpha="count")
    
)
save_fn = paste(save_dir, "/", "entropy_regression.png", sep="")
ggsave(
    save_fn,
    plot=h_regression,
    width=10,
    height=10,
)

# Regression for density and references
refs_regression <- (
  ggplot(
    df_zf,
    aes(x=density, y=references)
  )
  + geom_point(
    alpha=0.2,
  )
  + geom_smooth(color="orange", method="lm")
)
save_fn = paste(save_dir, "/", "references_regression.png", sep="")
ggsave(
  save_fn,
  plot=refs_regression,
  width=10,
  height=10,
)


metric_vs_year <- function(metric) {
  return(
    ggplot(
        df_zf,
        aes(
          x=year,
          y=.data[[metric]],
        )
      )
      + geom_point(alpha=0.2)
      + geom_smooth(color="orange", method="lm")
  )
}

for (metric in c("density", "edginess")) {
  # todo: refactor these into a func and call twice
    save_fn <- paste(save_dir, "/", metric, ".png", sep="")
    ggsave(
        save_fn,
        plot=metric_vs_cpy(metric),
        width=10,
        height=10,
    )

    save_fn <- paste(save_dir, "/", "year_vs_", metric, ".png", sep="")
      ggsave(
        save_fn,
        plot=metric_vs_year(metric),
        width=10,
        height=10,
    )
}

report <- paste("saved plots to", save_dir)
print(report)
